LanguageExt.Core/DataTypes/Seq/Enum.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace LanguageExt
+{
+    internal class Enum<A>
+    {
+        IEnumerator<A> iter;
+        List<A> list;
+
+        public Enum(IEnumerable<A> seq)
+        {
+            this.iter = seq.GetEnumerator();
+            this.list = new List<A>();
+        }
+
+        public Enum(IEnumerator<A> iter)
+        {
+            this.iter = iter;
+            this.list = new List<A>();
+        }
+
+        ~Enum()
+        {
+            iter?.Dispose();
+            iter = null;
+        }
+
+        public (bool Success, A Value) Get(int index)
+        {
+            if (index < list.Count)
```

Comment:
Is this safe if we're mid-growing? Will the `list.Count` be partially updated before we're finished and thus this could return uninitialized array element?  In other words, do we need to wrap in a `lock(list)`?

Reply:
`_size` in `List<T>` is always updated after the allocation if its internal array and any set operation.  So, even though it could be out of sync with the backing store, it will only ever report that the collection is smaller than it thinks.  So, the worst that could happen (I believe) is that we attempt to stream another item.   But there shouldn't be any synchronisation issues.

There's a unit test that pretty aggressively tests the lazy loading whilst multiple threads are trying to iterate it:

https://github.com/louthy/language-ext/blob/c412a962fac7110845b80d0f80a467cc929637ff/LanguageExt.Tests/Seq.Module.Tests.cs#L182